File: ElevationStatistics.m
% find study elevation range
%
% ElevationStatistics.m
% Created by Mac McMahan
% Last edited 4 Nov 2018
%
% Takes in data tables created by ArcMap Model and locates elevation bands
% which contain streams in all 3 rock types.  It first separates the files
% into matrices grouped by rock type.  From there hypsometric curves for
% each geographic group are calculated showing the overlap of the three
% rock types from each group.  The hypsometries are multiplied by each
% other within groups to locate only the 10 meter elevation band where all
% three hypsometries overlap.  The latitude and longitude of points which
% fall into those elevation bands are then extracted, along with new unique
% object IDs and stream segment IDs.  The points are then exported as .csv
% files to be loaded back into ArcMap for additional processing.
%
%--------------------------------------------------------------------------

clear;
clc;
tic; 

%reads name data files
data=ls('DataTables/*.xls');
%finds number of data files
[m,~]=size(data);
%fix bin edges, all histograms comparable, 4500 to acoomodate Mt. Whitney
edges=0:10:4500;

%---loop over calcareous---
%initialize counter, used for allocation into new, smaller matrices
count=1;
dcount=1;
%preallocates calcareous matrix, made big enough to accept most datasets
dataCalc=zeros(1000,m);
%preallocate hypsometry matrix, large enough to accomodate Mt. Whitney
%bin numbers range from 0 to 450, *10 to get elevation range 0 m to 4500 m
hypCalc=zeros(450,(m/3));
for i=1:3:m-2
    %path and name of data files to read in
    fileIn=sprintf('DataTables/%s',data(i,:));
    %load in data file
    excel=xlsread(fileIn);
    %find dimensions of data file
    [r,~]=size(excel);
    %load elevation column from data file into elevCalc, only fills the
    %necessary cells, others left unaffected
    dataCalc((1:r),(dcount:dcount+4))=excel(:,(1:5));
    %claclulates hypsometric data without plotting
    hypCalc(:,count)=histcounts(excel(:,5),edges);
    %increment counter
    count=count+1;
    dcount=dcount+5;
    
end

%---loop over felsic---
count=1;
dcount=1;
dataFels=zeros(1000,m);
hypFels=zeros(450,(m/3));
for j=2:3:m-1
    fileIn=sprintf('DataTables/%s',data(j,:));
    excel=xlsread(fileIn);
    [r,~]=size(excel);
    dataFels((1:r),(dcount:dcount+4))=excel(:,(1:5));
    hypFels(:,count)=histcounts(excel(:,5),edges);
    count=count+1;
    dcount=dcount+5;  
end

%---loop over mafic---
count=1;
dcount=1;
dataMaf=zeros(1000,m);
hypMaf=zeros(450,(m/3));
for k=3:3:m
    fileIn=sprintf('DataTables/%s',data(k,:));
    excel=xlsread(fileIn);
    [r,~]=size(excel);
    dataMaf((1:r),(dcount:dcount+4))=excel(:,(1:5));
    hypMaf(:,count)=histcounts(excel(:,5),edges);
    count=count+1;
    dcount=dcount+5;  
end

%---plot histograms---
for q=1:m/3
   figure;
   %plots elevation values that are not equal to zero
   histogram(dataCalc(dataCalc(:,q*5)~=0,q*5),'BinWidth',10);
   hold on;
   histogram(dataFels(dataFels(:,q*5)~=0,q*5),'BinWidth',10);
   histogram(dataMaf(dataMaf(:,q*5)~=0,q*5),'BinWidth',10);
   %give each plot a unique title
   txt=sprintf('Group %d hypsometries',q);
   title(txt);
   xlabel('Elevation (m)');
   ylabel('Occurence of elevation (count)');
   %give each figure a unique name
   fgnm=sprintf('Group%02dHypsometries.png',q);
   %path to output folder 
   path='ElevationBands/';
   saveas(gcf,fullfile(path,fgnm),'png');
   hold off;
end

%---locate hypsometric intersections---
%multiply hypsometry against each other, only 3way intersect(int) remains
int=hypCalc.*hypFels.*hypMaf;
%finds where intersection is not 0
elevRaw=(int(:,:)>0);

%find index of non-zero values from elevRaw
Calc=zeros(100,3);
Fels=zeros(100,3);
Maf=zeros(100,3);
for h=1:m/3
    %test to see if suitable conditions exist
    num=sum(int(:,h));
    if num==0
        fprintf('Group %d: No suitable locations found.\n',h);        
    else
        elev=find(elevRaw(:,h));
        %linear index range and convert to meters ASL
        MinE=min(elev)*10;
        MaxE=max(elev)*10;
        fprintf('Group %d: Suitable conditions between %d m and %d m elevation.\n',h,MinE,MaxE);
        
        %clears variables for new iteration
        clear Calc;
        clear Fels;
        clear Maf;
        
        %locates xyz data from dataROCKTYPE in elevation range from above
        Calc(:,(1:5))= dataCalc(dataCalc(:,(h*5))>=MinE & dataCalc(:,(h*5))<=MaxE,(h*5-4:h*5));
        %finds max ObjectID and StreamID from current iteration
        OBM=max(Calc(:,1));
        IDM=max(Calc(:,2));
        Fels(:,(1:5))= dataFels(dataFels(:,(h*5))>=MinE & dataFels(:,(h*5))<=MaxE,(h*5-4:h*5));
        %adds existing max OBJECTID and StreamID to all new values, maintains uniqueness
        Fels(:,1)=Fels(:,1)+OBM;
        Fels(:,2)=Fels(:,2)+IDM;
        %finds new max ObjectID and Stream ID
        OBM=max(Fels(:,1));
        IDM=max(Fels(:,2));
        Maf(:,(1:5))= dataMaf(dataMaf(:,(h*5))>=MinE & dataMaf(:,(h*5))<=MaxE,(h*5-4:h*5));
        %maintaining ID uniqueness by adding previous max to everything
        Maf(:,1)=Maf(:,1)+OBM;
        Maf(:,2)=Maf(:,2)+IDM;
        %loads all xyz points into single matrix for output
        Group=[Calc;Fels;Maf];
    
        %saves unique filename as a string
        fnm=sprintf('Group%02d.csv',h);
        fileOut=fullfile(path,fnm);
        %creates new file
        fid=fopen(fileOut,'w');
        %print header to file
        fprintf(fid,'ObjectID,StreamID,Long,Lat,Elevation\n');
        %append file with data
        dlmwrite(fileOut,Group,'delimiter',',','-append','precision','%16.12f');
        %close file
        fclose(fid);
    end
end

toc;

File: ElevationStatistics.py
# -*- coding: utf-8 -*-

"""
find study elevation range

ElevationStatistics.py
Created by William McMahan and Dr. William Armstrong
Last edited 9 Apr 2019

Takes in data tables created by QGIS Process and locates 10 m elevation bands
which contain streams in all 3 rock types.  It first separates the files
into matrices grouped by rock type.  From there hypsometric curves for
each geographic group are calculated showing the overlap of the three
rock types from each group.  The hypsometries are multiplied by each
other within groups to locate only the 10 meter elevation bands where all
three hypsometries overlap.  The latitude and longitude of points which
fall into that elevation range are then extracted, along with new unique
object IDs and stream segment IDs.  The points are then exported as .csv
files to be loaded back into QGIS for additional processing.
"""


#import necessary modules
from datetime import datetime
import os
import sys
import xlrd
import numpy as np
import matplotlib.pyplot as plt

startTime=datetime.now()

#read in names of data files
data=[]
for xls in os.listdir("DataTables/"):
    if xls.endswith(".xls"):
       data.append(xls) 

#number of data files located
numFiles=len(data)
#perform a check to make sure all files are present
if numFiles%3 != 0:
    #kills the script
    sys.exit()
#number of sites located
numSites=numFiles/3

#fix bin edges, all histograms comparable, 8850 to accommodate Mt. Everest
"""
using 4500 right now to make it simpler to work with
"""
edges=range(0,4510,10)


#---loop over carbonate---
#initialize matrices, 8850 to accommodate Mt. Everest
dataCarb=np.zeros((4500,numSites*5))
elevCarb=np.zeros((4500,numSites))
#this finds the index and file name for all calcareous files 
for index, item in enumerate(data):
    if index in range(0,numFiles,3):
        #saves the path to the data file as a string
        loc=("DataTables/%s" %(item))
        #opens the data file
        wb=xlrd.open_workbook(loc)
        #locates the data in the file
        sheet=wb.sheet_by_index(0)
        #determines the correct column index to place the data into
        datInd=index/3*5
        elvInd=index/3
        #read the data from the excel sheet and load it into matrix by column
        #i starts at 1 because of column headers in xls file
        for i in range(1,sheet.nrows):
        #following lines start at i-1 to place data in 0th index
            #objectID
            dataCarb[i-1,datInd]=sheet.cell_value(i,0)
            #streamID
            dataCarb[i-1,datInd+1]=sheet.cell_value(i,1)
            #Longitude
            dataCarb[i-1,datInd+2]=sheet.cell_value(i,2)
            #Latitude
            dataCarb[i-1,datInd+3]=sheet.cell_value(i,3)
            #Elevation
            dataCarb[i-1,datInd+4]=sheet.cell_value(i,4)
            #fills the separate elevation matrix
            elevCarb[i-1,elvInd]=sheet.cell_value(i,4)


#---loop over felsic---
dataFels=np.zeros((4500,numSites*5))
elevFels=np.zeros((4500,numSites))
for index, item in enumerate(data):
    if index in range(1,numFiles,3):
        loc=("DataTables/%s" %(item))
        wb=xlrd.open_workbook(loc)
        sheet=wb.sheet_by_index(0)
        datInd=index/3*5
        elvInd=index/3
        for i in range(1,sheet.nrows):
            dataFels[i-1,datInd]=sheet.cell_value(i,0)
            dataFels[i-1,datInd+1]=sheet.cell_value(i,1)
            dataFels[i-1,datInd+2]=sheet.cell_value(i,2)
            dataFels[i-1,datInd+3]=sheet.cell_value(i,3)
            dataFels[i-1,datInd+4]=sheet.cell_value(i,4) 
            elevFels[i-1,elvInd]=sheet.cell_value(i,4) 
    
    
#---loop over mafic---
dataMaf=np.zeros((4500,numSites*5))
elevMaf=np.zeros((4500,numSites))
for index, item in enumerate(data):
    if index in range(2,numFiles,3):
        loc=("DataTables/%s" %(item))
        wb=xlrd.open_workbook(loc)
        sheet=wb.sheet_by_index(0)
        datInd=index/3*5
        elvInd=index/3
        for i in range(1,sheet.nrows):
            dataMaf[i-1,datInd]=sheet.cell_value(i,0)
            dataMaf[i-1,datInd+1]=sheet.cell_value(i,1)
            dataMaf[i-1,datInd+2]=sheet.cell_value(i,2)
            dataMaf[i-1,datInd+3]=sheet.cell_value(i,3)
            dataMaf[i-1,datInd+4]=sheet.cell_value(i,4)
            elevMaf[i-1,elvInd]=sheet.cell_value(i,4) 
 
   
#---plot histograms and calculate hypsometries---
#set all 0 values to NaN for plotting
elevCarb[elevCarb==0]=np.nan
elevFels[elevFels==0]=np.nan
elevMaf[elevMaf==0]=np.nan
#preallocate hypsometry matrices
hypCarb=np.zeros((450,numSites))
hypFels=np.zeros((450,numSites))
hypMaf=np.zeros((450,numSites))
#iterates over the number of sites returned from map analysis
for site in range(numSites):
    
    #--plot histograms--
    #changes 0th index for site naming
    nm=site+1
    #saves the site name as a string
    name=("Site %d" %nm)
    #plots the histograms, bins range from the integer of the lowest elevation
    #in the dataset to 10m higher than the integer of the highest elevation 
    #in the dataset, with bins 10m wide.
    plt.hist(elevCarb[:,site],color="b",alpha=0.3,bins=range(int(min(elevCarb[:,site])),int(max(elevCarb[:,site])+10),10))
    plt.hist(elevFels[:,site],color="r",alpha=0.3,bins=range(int(min(elevFels[:,site])),int(max(elevFels[:,site])+10),10))
    plt.hist(elevMaf[:,site],color="g",alpha=0.3,bins=range(int(min(elevMaf[:,site])),int(max(elevMaf[:,site])+10),10))
    #palces a legend in the upper right
    plt.legend(("Carbonate","Felsic","Mafic"),loc=1)
    #title the plot
    plt.title(name)
    #label the axes
    plt.xlabel("Elevation (m ASL)")
    plt.ylabel("Count")
    #draw all 3 rock types on single plot
    plt.show()     
    
    #--calculate hypsometries--
    #calculates distribution of points into specified elevation bands
    binCarb=np.histogram(elevCarb[:,site],bins=edges)
    #saves tuple value from each loop iteration into single array
    hypCarb[:,site]=binCarb[0]
    
    binFels=np.histogram(elevFels[:,site],bins=edges)
    hypFels[:,site]=binFels[0]
    """
    hypMaf seems to be where the issue is.  The non-zero numbers are in bins
    38-43, when they should be in bins 48-68.  They are in the correct place 
    for every other rock type and iteration, it is just hypMaf in the second
    iteration (site==1) that seems to be raising an issue.
    """
    binMaf=np.histogram(elevMaf[:,site],bins=edges)
    hypMaf[:,site]=binMaf[0]
    

#---locate hypsometric intersection---
#multiply hypsometries against each other, only 3way intersect remains
intsect=np.multiply(hypCarb,hypFels,hypMaf)
"""
The following line of hard code fixes the issue, but will prevent any site 2 
from ever having suitable elevations, even if it really does work.  This line 
will need to be removed by fixing the issue above
"""
intsect[:,1]=np.zeros((450))
#find where value is not 0
elevRaw=intsect>0


#---output information to user---
for i in range(numSites):
    #convoluted equation to equate input and output table columns
    j=((i+1)*5)-1
    #sum column corresponding to site number
    num=np.sum(elevRaw[:,i])
    if num==0:
        print("Group %d: No suitable locations found." %(i+1))
    else:
        #locates indeces of non-zero values
        elev=np.where(elevRaw[:,i])
        #find max and min elevations in meters ASL
        minE=np.min(elev)*10
        maxE=np.max(elev)*10
        print("Group %d: Suitable conditions between %d m and %d m elevation." %((i+1),minE,maxE))
        
        #allocate matrices and clear from last iteration
        Carb=np.zeros((100,5))
        Fels=np.zeros((100,5))
        Maf=np.zeros((100,5))

        #extract all rows of data where elevation is within range
        dumpC=dataCarb[(dataCarb[:,j]>=minE) & (dataCarb[:,j]<=maxE)]
        #save only the columns corresponding to the currently analyzed site
        Carb=dumpC[:,i*5:j+1]
        #finds max ObjectID and StreamID from the current iteration
        OBM=np.max(Carb[:,0])
        IDM=np.max(Carb[:,1])
        dumpF=dataFels[(dataFels[:,j]>=minE) & (dataFels[:,j]<=maxE)]
        Fels=dumpF[:,i*5:j+1]
        #adds existing IDs to all new values, maintains uniqueness
        Fels[:,0]=Fels[:,0]+OBM
        Fels[:,1]=Fels[:,1]+IDM
        #finds new max ObjectID and StreamID
        OBM=np.max(Fels[:,0])
        IDM=np.max(Fels[:,1])
        dumpM=dataMaf[(dataMaf[:,j]>=minE) & (dataMaf[:,j]<=maxE)]
        Maf=dumpM[:,i*5:j+1]
        #maintains ID uniqueness
        Maf[:,0]=Maf[:,0]+OBM
        Maf[:,1]=Maf[:,1]+IDM
        #vertically concatenates all 3 rock type matrices into a group matrix
        Group=np.vstack((Carb,Fels,Maf))
        #save output file name and path
        outName=("ElevationBands/Group%02d.csv" %(i+1))
        #writes the output as a .csv to the output folder
        np.savetxt(outName,Group,header="ObjectID,StreamID,Long,Lat,Elevation",delimiter=",")
        
        
print(datetime.now()-startTime)    